Docs: Update the docs for UI-Extensions (#881)
grafana/plugin-tools: docusaurus/docs/ui-extensions/create-an-extension-point.md

  - apps
---

> ## What is an extension point?

Reviewer: When the H1 (# Create an extension point) is added above this H2, it creates stacked headers which Grafana docs try to avoid. I suggest no H2  or adding a sentence above the H2 to introduce the topic as a whole.

Author: Thanks, that's a good point, I wasn't aware of this unfortunately 🙏 
I'll go with [Jack's suggestion](https://github.com/grafana/plugin-tools/pull/881#discussion_r1574263489) which adds a sentence above the H2, let me know what you think.

> An extension point is a place in the UI which you make extendable by other plugins using extensions. These extensions can be either be links or simple React components (that can basically impelement anything). It's up to you how you are displaying the handling or displaying the extensions, we have some examples below.

Reviewer: ```suggestion
An extension point is a place in the UI which you make extendable by other plugins using extensions. These extensions can either be links or React components that can impelement virtually anything. It's up to you how to display the handling or displaying of extensions. Refer to the following examples below:
```

Author: Hmm, should we keep `following` and the `:` in the last line? (I am asking because the examples are not directly following this sentence, but they are a few paragraphs below.) Maybe we should add a link instead?

> You can also share contextual information with the extensions using a `context` object or component props (see the examples below).

Reviewer: ```suggestion
You can also share contextual information with the extensions using a `context` object or component props (see the following examples).
```

### Requirements

- **an extension point ID** _(string)_
  - In case it's an extension point in core Grafana, it must start with `grafana/`
  - In case it's inside a plugin, it must start with `plugin/<PLUGIN_ID>/`
  - It must be unique

> - **an app plugin** - _currently only app plugins can create extension points_

Reviewer: ```suggestion
- App plugins - _currently only app plugins can create extension points_
```

Author: Hmm, I am not sure about this one - but maybe I am wrong. When we are listing the requirements, wouldn't `an app plugin` or `app plugin` sounds better than `App plugins` (in plural)?

:::note

> Consider the design the UI of the extension point so it supports a scenario where multiple extensions are being added without breaking the UI. Also consider if there is any information from the current view that should be shared with the extensions added to the extension point. It could be information from the current view that could let the extending plugin prefill values or other data in the functionality being added via the extension.

Reviewer: ```suggestion
Consider the design of the UI of the extension point so it supports a scenario where multiple extensions can be added without breaking the UI. Also consider if there is any information from the current view that should be shared with the extensions added to the extension point. It could be information from the current view that could let the extending plugin prefill values or other data in the extension's functionality.
```

Reviewer: Grafana docs prefer sentences < 25 words in length.

:::

## Create an extension point

:::danger

When you create an extension point in a plugin, you create a public interface for other plugins to interact with. Changes to the extension point ID or its context could break any plugin that attempts to register a link inside your plugin.

:::

You can easily create an extension point using the following functions (they live in `@grafana/runtime`) to fetch extensions for a certain extension point ID:

> ### `usePluginExtensions()`

Reviewer: ```suggestion
### The `usePluginExtensions()` hook
```

> :::note

Reviewer: ```tip
:::note
```

Author: I'll resolve this - let me know if you would like any changes in this note block.

> **Performance tip:** in case the `context` object is created dynamically, make sure to wrap it into a `useMemo()` to prevent unnecesssary rerenders. [More info](#)

Reviewer: ```suggestion
In case the `context` object is created dynamically, make sure to wrap it into a `useMemo()` to prevent unnecesssary rerenders. [More info](#)
```

:::

> The `usePluginExtensions` react hooks returns the list of extensions that are registered for a certain extension point ID. The hook will dynamically update its return value when the list of extensions changes - this usually happens when extensions are registered during runtime due to dynamic plugin loading.

Reviewer: ```suggestion
The `usePluginExtensions` React hooks returns the list of extensions that are registered for a certain extension point ID. The hook dynamically updates its return value when the list of extensions changes. This behavior usually happens when extensions are registered during runtime due to dynamic plugin loading.
```

#### Syntax

```tsx
usePluginExtensions(options);
usePluginLinkExtensions(options); // Only returns extensions that have type `type="link"`
usePluginComponentExtensions(options); // Only returns extensions that have type `type="component"`
```

#### Parameters

##### `options.extensionPointId` - _string_

> The unique identifier of your extension point. It must begin with `plugins/<PLUGIN_ID>`, for example: `plugins/myorg-super-app`.

Reviewer: ```suggestion
The unique identifier of your extension point. It must begin with `plugins/<PLUGIN_ID>`. For example: `plugins/myorg-super-app`.
```

##### `options?.context` - _object (Optional)_

> An arbitrary object, that contains information related to your extension point which you would like to share with the extensions, for example: `{ baseUrl: '/foo/bar' }`. This parameter is not available for component extensions, for those you can pass contextual information using the component props. **Note:** the provided context object always gets frozen (turned immutable) before being shared with the extensions.

Reviewer: ```suggestion
As an arbitrary object, it contains information related to your extension point that you would like to share with the extensions. For example: `{ baseUrl: '/foo/bar' }`. This parameter is not available for component extensions because you can pass contextual information using the component props. 

:::note

The provided context object always gets frozen (turned immutable) before being shared with the extensions.

:::
```

##### `options?.limitPerPlugin` - _number (Optional)_

> It can be used to limit maximum how many extensions should be returned from the same plugin. It can be useful in cases when there is limited space on the UI to display extensions.

Reviewer: ```suggestion
Use thie parameter to set the maximum value for how many extensions should be returned from the same plugin. It can be useful in cases when there is limited space on the UI to display extensions.
```

#### Return value

The hooks return an object in the following format:

```tsx
// usePluginExtensions()
{
  isLoading: boolean;
  extensions: Array<PluginExtensionLink | PluginExtensionComponent>
}

// usePluginLinkExtensions()
{
  isLoading: boolean;
  extensions: PluginExtensionLink[]
}

// usePluginComponentExtensions()
{
  isLoading: boolean;
  extensions: PluginExtensionComponent[]
}
```

(For more information check the type definitions of [`PluginExtensionLink`](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/pluginExtensions.ts#L27) and [`PluginExtensionComponent`](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/pluginExtensions.ts#L35C13-L35C37).)

#### Example - rendering link extensions (static context)

> In the following example we render a link component all link-type extensions that other plugins registered for the `plugins/another-app-plugin/menu` extension point ID.

Reviewer: ```suggestion
The following example shows how to render a link component all link-type extensions that other plugins registered for the `plugins/another-app-plugin/menu` extension point ID.
```

Reviewer: The above sentence is unclear ("render a link component all link-type...") but I'm not sure how to fix it.

```tsx
import { usePluginLinkExtensions } from '@grafana/runtime';

// We are defining the `context` outside of the React component for performance reasons.
// (In case we would declare it inside the component below, then it would result in a new object on every render,
// which would unnecessarily trigger the `usePluginLinkExtensions()` hook.)
const context = {
  referenceId: '12345',
  timeZone: 'UTC',
};

function AppMenuExtensionPoint() {
  // This only returns type="link" extensions
  const { extensions } = usePluginLinkExtensions({
    extensionPointId: 'plugins/another-app-plugin/menu',
    context,
  });

  if (extensions.length === 0) {
    return null;
  }

  return (
    <div>
      {extensions.map((extension) => {
        return (
          <a href={extension.path} onClick={extension.onClick} title={extension.description} key={extension.key}>
            {extension.title}
          </a>
        );
      })}
    </div>
  );
}
```

#### Example - rendering link extensions (dynamic context)

> In this example we are creating the context object dynamically. Although this is a common practice, we have to be aware, that the `usePluginLinkExtensions()` hook will rerender in the following scenarios:

Reviewer: ```suggestion
The following example shows how to create the context object dynamically. Although this is a common practice, you should be aware that the `usePluginLinkExtensions()` hook will re-render in the following scenarios:
```

> - IF the `context` object changes _(so the extensions can react to the context changes)_

Reviewer: ```suggestion
- If the `context` object changes _(so the extensions can react to the context changes)_
```

> - IF the extension-registry changes

Reviewer: ```suggestion
- If the extension-registry changes
```

Reviewer: Only italics for emphasis if absolutely necessary.

> Because of these we have to make sure, that we only change the `context` object if its content changes, otherwise we will cause unnecessary rerenders. The following example shows how to approach these scenarios in a safe way:

Reviewer: ```suggestion
Be sure to only change the `context` object if its content changes; otherwise, you could create unnecessary re-renders. The following example shows how to approach these scenarios in a safe way:
```

```tsx
import { useMemo } from 'react';
import { usePluginLinkExtensions } from '@grafana/runtime';

function AppMenuExtensionPoint({ referenceId }) {
  // Instead of just defining the object here (which would result in a new object on every render),
  // we use `useMemo()` to only update the context object when its "dynamic" dependencies change.
  const context = useMemo(
    () => ({
      referenceId,
      timeZone: 'UTC',
    }),
    [referenceId]
  );
  const { extensions } = usePluginLinkExtensions({
    extensionPointId: 'plugins/another-app-plugin/menu',
    context,
  });

  if (extensions.length === 0) {
    return null;
  }

  return (
    <div>
      {extensions.map((extension) => {
        return (
          <a href={extension.path} onClick={extension.onClick} title={extension.description} key={extension.key}>
            {extension.title}
          </a>
        );
      })}
    </div>
  );
}
```

#### Example - rendering component extensions

:::note

> **Available in Grafana >=10.1.0** <br /> (_Component type extensions are only available in Grafana 10.1.0 and above._)

Reviewer: ```suggestion
**This feature is only available in Grafana versions 10.1.0 and above.
```

:::

> Component type extensions are simple React components, which gives much more freedom on what they are able to do. You can pass contextual information to the extension components using props.

Reviewer: ```suggestion
Component type extensions are simple React components, which gives you much more freedom in  what you can make them do. For example, you can pass contextual information to the extension components using props.
```

Author: Thanks! I have fixed it in https://github.com/grafana/plugin-tools/pull/881/commits/a5821d379642c396fdd2163dcecb71afb89a7e9c, however I left out the `For example` part (I think this is more like an additional information than an example), please let me know if you disagree.

```tsx title="src/components/Toolbar.tsx"
import { usePluginComponentExtensions } from '@grafana/runtime';

export const Toolbar = () => {
  // This only returns type="component" extensions
  // Heads up! We don't specify a context object below, we pass in the contextual information as a prop to the component later.
  const { extensions } = usePluginComponentExtensions({ extensionPointId: '<extension-point-id>' });

  return (
    <div>
      <div className="title">Title</div>
      <div className="extensions">
        {/* Loop through the available extensions */}
        {extensions.map((extension) => {
          const Component = extension.component as React.ComponentType<{
            version: string;
          }>;

          // Render extension component and pass contextual information (version)
          return (
            <div key={extension.id}>
              <Component version="1.0.0" />
            </div>
          );
        })}
      </div>
    </div>
  );
};
```

> ### `getPluginExtensions()` - _deprecated_

Reviewer: ```suggestion
### The `getPluginExtensions()` method - _deprecated_
```

The `getPluginExtensions` method takes an object consisting of the `extensionPointId`, which must begin `plugins/<PLUGIN_ID>`, and any contextual information that you want to provide. The `getPluginLinkExtensions` method returns a list of extension links that your program can then loop over.

:::note

This function only returns the state of the extensions registry (the extensions registered by plugins) at a given time. If there are extensions registered by plugins after that point in time, you won't receive them. <br />

> **We strongly suggest to use the reactive [`usePluginExtensions()`](#usepluginextensions) hook instead wherever possible.**

Reviewer: ```suggestion
As a best practice, use the reactive [`usePluginExtensions()`](#usepluginextensions) hook instead wherever possible.
```

:::

#### Syntax

```tsx
getPluginExtensions(options);
getPluginLinkExtensions(options); // Only returns extensions that have type `type="link"`
getPluginComponentExtensions(options); // Only returns extensions that have type `type="component"`
```

#### Parameters

##### `options.extensionPointId` - _string_

> The unique identifier of your extension point. It must begin with `plugins/<PLUGIN_ID>`, for example: `plugins/myorg-super-app`.

Reviewer: ```suggestion
The unique identifier of your extension point. It must begin with `plugins/<PLUGIN_ID>`. For example: `plugins/myorg-super-app`.
```

##### `options?.context` - _object (Optional)_

> An arbitrary object, that contains information related to your extension point which you would like to share with the extensions, for example: `{ baseUrl: '/foo/bar' }`. This parameter is not available for component extensions, for those you can pass contextual information using the component props. **Note:** the provided context object always gets frozen (turned immutable) before being shared with the extensions.

Reviewer: ```suggestion
As an arbitrary object, it contains information related to your extension point which you would like to share with the extensions. For example: `{ baseUrl: '/foo/bar' }`. 

This parameter is not available for component extensions; for those, you can pass contextual information using the component props. 

:::note

The provided context object always gets frozen (turned immutable) before being shared with the extensions.

:::
```

> ##### `options?.limitPerPlugin` - _number (Optional)_

Reviewer: ```suggestion
##### The `options?.limitPerPlugin` method - _number (Optional)_
```

Author: I guess this suggestion is maybe a typo? (This is one of the option properties). I will resolve it, please reopen if you have objections!

> It can be used to limit maximum how many extensions should be returned from the same plugin. It can be useful in cases when there is limited space on the UI to display extensions.

Reviewer: ```suggestion
Use this method to specify the maximum amount of extensions that should be returned from the same plugin. It can be useful in cases when there is limited space on the UI to display extensions.
```

  - apps
---

> ## What is an extension point?

Reviewer 2: I agree with Joe here. It's a bit funny to have stacked headers. Maybe we can do something like:

```suggestion

In this guide you will learn how to provide an extension point so that other app plugins can add their extensions to your plugin.

## What is an extension point?
```

An extension point is a place in the UI which you make extendable by other plugins using extensions. These extensions can be either be links or simple React components (that can basically impelement anything). It's up to you how you are displaying the handling or displaying the extensions, we have some examples below.

You can also share contextual information with the extensions using a `context` object or component props (see the following examples).

### Requirements

- **an extension point ID** _(string)_
  - In case it's an extension point in core Grafana, it must start with `grafana/`
  - In case it's inside a plugin, it must start with `plugin/<PLUGIN_ID>/`
  - It must be unique
- **an app plugin** - _currently only app plugins can create extension points_

:::note

> Consider the design of the UI of the extension point so it supports a scenario where multiple extensions can be added without breaking the UI. Also consider if there is any information from the current view that should be shared with the extensions added to the extension point. It could be information from the current view that could let the extending plugin prefill values or other data in the extension's functionality.

Reviewer 2: ```suggestion
When designing the UI make sure the extension point supports a scenario where multiple extensions can be added without breaking the UI. Also consider if there is any information from the current view that should be shared with the extensions added to the extension point. It could be information from the current view that could let the extending plugin pre-fill values or other data in the extension's functionality.
```

:::

You can easily create an extension point using the following functions (they live in `@grafana/runtime`) to fetch extensions for a certain extension point ID:

### The `usePluginExtensions()` hook

:::note

In case the `context` object is created dynamically, make sure to wrap it into a `useMemo()` to prevent unnecesssary rerenders. [More info](#)

:::

The `usePluginExtensions` React hooks returns the list of extensions that are registered for a certain extension point ID. The hook dynamically updates its return value when the list of extensions changes. This behavior usually happens when extensions are registered during runtime due to dynamic plugin loading.

#### Syntax

```tsx
usePluginExtensions(options);
usePluginLinkExtensions(options); // Only returns extensions that have type `type="link"`
usePluginComponentExtensions(options); // Only returns extensions that have type `type="component"`
```

#### Parameters

##### `options.extensionPointId` - _string_

The unique identifier of your extension point. It must begin with `plugins/<PLUGIN_ID>`. For example: `plugins/myorg-super-app`.

##### `options?.context` - _object (Optional)_

> As an arbitrary object, it contains information related to your extension point that you would like to share with the extensions. For example: `{ baseUrl: '/foo/bar' }`. This parameter is not available for component extensions because you can pass contextual information using the component props.

Reviewer 2: ```suggestion
An object containing information related to your extension point that you would like to share with the extensions. For example: `{ baseUrl: '/foo/bar' }`. This parameter is not available for component extensions, instead you should pass contextual information using the component props.
```

:::note

The provided context object always gets frozen (turned immutable) before being shared with the extensions.

:::

##### `options?.limitPerPlugin` - _number (Optional)_

> Use thie parameter to set the maximum value for how many extensions should be returned from the same plugin. It can be useful in cases when there is limited space on the UI to display extensions.

Reviewer 2: ```suggestion
Use this parameter to set the maximum value for how many extensions should be returned from the same plugin. It can be useful in cases when there is limited space on the UI to display extensions.
```

#### Return value

The hooks return an object in the following format:

```tsx
// usePluginExtensions()
{
  isLoading: boolean;
  extensions: Array<PluginExtensionLink | PluginExtensionComponent>
}

// usePluginLinkExtensions()
{
  isLoading: boolean;
  extensions: PluginExtensionLink[]
}

// usePluginComponentExtensions()
{
  isLoading: boolean;
  extensions: PluginExtensionComponent[]
}
```

(For more information check the type definitions of [`PluginExtensionLink`](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/pluginExtensions.ts#L27) and [`PluginExtensionComponent`](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/pluginExtensions.ts#L35C13-L35C37).)

#### Example - rendering link extensions (static context)

> The following example shows how to render a link component all link-type extensions that other plugins registered for the `plugins/another-app-plugin/menu` extension point ID.

Reviewer 2: ```suggestion
The following example shows how to render a link component as link-type extensions that other plugins registered for the `plugins/another-app-plugin/menu` extension point ID.
```

```tsx
import { usePluginLinkExtensions } from '@grafana/runtime';

> // We are defining the `context` outside of the React component for performance reasons.
> // (In case we would declare it inside the component below, then it would result in a new object on every render,
> // which would unnecessarily trigger the `usePluginLinkExtensions()` hook.)

Reviewer 2: ```suggestion
// We define the `context` outside of the React component for performance reasons.
// (Declaring it inside the component would result in a new object on every render,
// which would unnecessarily trigger the `usePluginLinkExtensions()` hook.)
```

    extensionPointId: 'plugins/another-app-plugin/menu',
    context,
  });

  if (extensions.length === 0) {
    return null;
  }

  return (
    <div>
      {extensions.map((extension) => {
        return (
          <a href={extension.path} onClick={extension.onClick} title={extension.description} key={extension.key}>
            {extension.title}
          </a>
        );
      })}
    </div>
  );
}
```

#### Example - rendering link extensions (dynamic context)

The following example shows how to create the context object dynamically. Although this is a common practice, you should be aware that the `usePluginLinkExtensions()` hook will re-render in the following scenarios:

- If the `context` object changes _(so the extensions can react to the context changes)_
- If the extension-registry changes

Be sure to only change the `context` object if its content changes; otherwise, you could create unnecessary re-renders. The following example shows how to approach these scenarios in a safe way:

```tsx
import { useMemo } from 'react';
import { usePluginLinkExtensions } from '@grafana/runtime';

function AppMenuExtensionPoint({ referenceId }) {

>   // Instead of just defining the object here (which would result in a new object on every render),
>   // we use `useMemo()` to only update the context object when its "dynamic" dependencies change.

Reviewer 2: ```suggestion
  // Instead of defining the object here (which would result in a new object on every render),
  // we use `useMemo()` to only update the context object when its "dynamic" dependencies change.
```

  const context = useMemo(
    () => ({
      referenceId,
      timeZone: 'UTC',
    }),
    [referenceId]
  );
  const { extensions } = usePluginLinkExtensions({
    extensionPointId: 'plugins/another-app-plugin/menu',
    context,
  });

  if (extensions.length === 0) {
    return null;
  }

  return (
    <div>
      {extensions.map((extension) => {
        return (
          <a href={extension.path} onClick={extension.onClick} title={extension.description} key={extension.key}>
            {extension.title}
          </a>
        );
      })}
    </div>
  );
}
```

#### Example - rendering component extensions

:::note

> \*\*This feature is only available in Grafana versions 10.1.0 and above.

Reviewer 2: Is this true? The example below uses `usePluginComponentExtensions` which I don't think is available in 10.1.0. 🤔

Author: I think I'll remove this note for now - unfortunately it's not yet sure which official release the feature will land in.

  - apps
---

> In this guide you will learn how to provide an extension point so that other app plugins can add their extensions to your plugin.

Reviewer 3: ```suggestion
In this guide you will learn how to provide an extension point so that app plugins can add their extensions to your plugin.
```
nit: if the page targets Core developers

Author: Good one 👍

## What is an extension point?

> An extension point is a place in the UI which you make extendable by other plugins using extensions. These extensions can be either be links or simple React components (that can basically impelement anything). It's up to you how you are displaying the handling or displaying the extensions, we have some examples below.

Reviewer 3: I'm not 100% sure we need to define it _again_, but not a blocking comment https://grafana.com/developers/plugin-tools/ui-extensions/#definition-of-terms

Building on @josmperez 's suggestion above and incorporating your response @leventebalogh 
```suggestion
An extension point is a place in the UI which you make extendable by other plugins using extensions. These extensions can either be links or React components that can implement virtually anything. 

In the extension point you can control how you want to handle the displaying of extensions to users. You can also share contextual information with the extensions using a `context` object or component props. Refer to the examples below to learn more.
```
An Examples subsection might be useful to break up the content and provide us a way to link from here. Note - I've also moved the content from below as we had two sections both ending with "see the examples"

You can also share contextual information with the extensions using a `context` object or component props (see the following examples).

> ### Requirements

Reviewer 3: Comment - requirements feels odd to me, we'd typically use this for things like Go version etc, there's other points below which might change how we want to title this section - can follow up.

- **an extension point ID** _(string)_
  - In case it's an extension point in core Grafana, it must start with `grafana/`
  - In case it's inside a plugin, it must start with `plugin/<PLUGIN_ID>/`
  - It must be unique

> - **an app plugin** - _currently only app plugins can create extension points_

Reviewer 3: What about Core developers who we'd direct this to as how to implement an extension point?

:::note

When designing the UI make sure the extension point supports a scenario where multiple extensions can be added without breaking the UI. Also consider if there is any information from the current view that should be shared with the extensions added to the extension point. It could be information from the current view that could let the extending plugin pre-fill values or other data in the extension's functionality.

:::

## Create an extension point

:::danger

> When you create an extension point in a plugin, you create a public interface for other plugins to interact with. Changes to the extension point ID or its context could break any plugin that attempts to register a link inside your plugin.

Reviewer 3: The "Note" feels odd in the requirements, and the "Danger" is a bit off-putting as the start of the "Create an extension point" section.

What about a section like "Considerations for building an extension point" which could detail these points? Open to suggestions on alternative name

Author: I actually like that idea, was thinking something similar (e.g. "Best practices"). I would probably go forward with this PR like this (to support the just-about-to-be-merged reactive extensions registry with docs), and update in an upcoming PR if that's fine with you.

:::

> You can easily create an extension point using the following functions (they live in `@grafana/runtime`) to fetch extensions for a certain extension point ID:

Reviewer 3: ```suggestion
You can easily create an extension point using the following functions from `@grafana/runtime` to fetch extensions for a certain extension point ID:
```
Avoids the use of brackets

Reviewer 3: Its also somewhat brief as an introduction. "Create an extension point" sounds like I'm going to get a tutorial, but instead we've got more of an API doc with examples. Additionally we say functions plural, but then have 2 "sections" with one of them a deprecated method.

I don't object to the format, but I think we may need to expand this sentence. Happy to follow up with more suggestions if you and agree and would like some help with this

Author: Yeah, we could probably add more content here, let's try to make it better together 👍 

> but instead we've got more of an API doc with examples

True, although maybe if we start reorganising the structure of the docs soon ("Reference" and How to ..." are separate sections), then it will be good for the API reference.

> Additionally we say functions plural, but then have 2 "sections" with one of them a deprecated method.

Hmm, I guess as long as we don't remove the getter functions plural is legit?

### The `usePluginExtensions()` hook

:::note

> In case the `context` object is created dynamically, make sure to wrap it into a `useMemo()` to prevent unnecesssary rerenders. [More info](#)

Reviewer 3: > In case the `context` object is created dynamically,

Why would this be the case? Would describing an example help here?

Author: This would happen when the `context {}` object is constructed dynamically from other variables, which can change between renders. Strange that Github didn't update the markdown snippet above, as the I have linked it to an example 🤔 (Probably I did that after your comment?)

```
[More info](#example---rendering-link-extensions-dynamic-context)
```

:::

> The `usePluginExtensions` React hooks returns the list of extensions that are registered for a certain extension point ID. The hook dynamically updates its return value when the list of extensions changes. This behavior usually happens when extensions are registered during runtime due to dynamic plugin loading.

Reviewer 3: ```suggestion
The `usePluginExtensions` React hooks return the list of extensions that are registered for a certain extension point ID. The hook dynamically updates its return value when the list of extensions changes. This behavior usually happens when extensions are registered during runtime due to dynamic plugin loading.
```

I'm not sure what's correct here... but if `hooks` is plural, return should be singular. The reason i'm not sure is that the section is `hook` singular, but the syntax lists 3 functions.

Author: No no, your suggestion is spot on (I meant talking about the three hooks, I just didn't want to type them out for making reading easier).

#### Syntax

```tsx
usePluginExtensions(options);
usePluginLinkExtensions(options); // Only returns extensions that have type `type="link"`

> usePluginComponentExtensions(options); // Only returns extensions that have type `type="component"`

Reviewer 3: I personally find it odd that these are all under "usePluginExtensions()" subsection but I may be in the minority.

Author: Yeah, maybe you are right. I only put them here for now as their syntax, arguments and return values are basically the same and I wanted to avoid too much duplication. On the other hand if we move towards a direction of having an API reference, then we will need to separate them, you are right.

```

#### Parameters

##### `options.extensionPointId` - _string_

> The unique identifier of your extension point. It must begin with `plugins/<PLUGIN_ID>`. For example: `plugins/myorg-super-app`.

Reviewer 3: what if Core?

Author: Good point, pushed a commit to fix it. 👍

##### `options?.context` - _object (Optional)_

> An object containing information related to your extension point that you would like to share with the extensions. For example: `{ baseUrl: '/foo/bar' }`. This parameter is not available for component extensions, instead you should pass contextual information using the component props.

Reviewer 3: > instead you should pass contextual information using the component props

We make this statement a lot... so a few questions:
- should we be providing advice to duplicate contextual information in both manners? If that is the case, why is it on the extension point developer? My worry is that it is up to the _extension_ developer to decide whether theres is a link or component, but by extending the same extension point they may get different context available, feels confusing.
- We reference it in the `Example - rendering component extensions` but maybe it is worth having an even smaller snippet which we could then link to from any time we make this statement?

Author: > We reference it in the Example - rendering component extensions but maybe it is worth having an even smaller snippet which we could then link to from any time we make this statement?

Yeah, I think that's a good idea! How about creating that sub-section "Considerations..." or "Best practices" and adding this as a point under there?

> should we be providing advice to duplicate contextual information in both manners? If that is the case, why is it on the extension point developer? My worry is that it is up to the extension developer to decide whether theres is a link or component, but by extending the same extension point they may get different context available, feels confusing.

Hmm, I don't think so, the two context objects should be the same, the difference is in how the extensions receive it. Probably the communication of this is not clear though, I think we should have a chat about it and try to make sure it's easier to understand.

:::note

> The provided context object always gets frozen (turned immutable) before being shared with the extensions.

Reviewer 3: ```suggestion
The provided context object is made immutable before being shared with the extensions.
```

:::

##### `options?.limitPerPlugin` - _number (Optional)_

> Use this parameter to set the maximum value for how many extensions should be returned from the same plugin. It can be useful in cases when there is limited space on the UI to display extensions.

Reviewer 3: It might be good to link to this from the section regarding designing your UI to respond to multiple plugins? But I think we also need to describe how it would be determined which plugins get displayed when there are more extensions than the limit allows

Author: Yeah, good point, and some more explanation would be good indeed. I will leave it for a separate refinement PR if you don't mind, let me know if you have objections.

#### Return value

The hooks return an object in the following format:

```tsx
// usePluginExtensions()
{
  isLoading: boolean;
  extensions: Array<PluginExtensionLink | PluginExtensionComponent>
}

// usePluginLinkExtensions()
{
  isLoading: boolean;
  extensions: PluginExtensionLink[]
}

// usePluginComponentExtensions()
{
  isLoading: boolean;
  extensions: PluginExtensionComponent[]
}
```

(For more information check the type definitions of [`PluginExtensionLink`](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/pluginExtensions.ts#L27) and [`PluginExtensionComponent`](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/pluginExtensions.ts#L35C13-L35C37).)

#### Example - rendering link extensions (static context)

The following example shows how to render a link component as link-type extensions that other plugins registered for the `plugins/another-app-plugin/menu` extension point ID.

```tsx
import { usePluginLinkExtensions } from '@grafana/runtime';

// We define the `context` outside of the React component for performance reasons.
// (Declaring it inside the component would result in a new object on every render,

> // which would unnecessarily trigger the `usePluginLinkExtensions()` hook.)

Reviewer 3: Do we need this detail in the docs?

Author: Yes, I think we do, it can be important help for plugin devs. (On the other this could be another candidate for having a "Considerations (...)" / "Best practices" section.)

const context = {
  referenceId: '12345',
  timeZone: 'UTC',
};

function AppMenuExtensionPoint() {
  // This only returns type="link" extensions
  const { extensions } = usePluginLinkExtensions({
    extensionPointId: 'plugins/another-app-plugin/menu',
    context,
  });

  if (extensions.length === 0) {
    return null;
  }

  return (
    <div>
      {extensions.map((extension) => {
        return (
          <a href={extension.path} onClick={extension.onClick} title={extension.description} key={extension.key}>
            {extension.title}
          </a>
        );
      })}
    </div>
  );
}
```

> #### Example - rendering link extensions (dynamic context)

Reviewer 3: I still find static vs dynamic a little unclear as to why the different situations occur.

Author: Good feedback, let's discuss how the best we could explain this difference in the docs.

:::

Component type extensions are simple React components, which gives you much more freedom in what you can make them do. You can pass contextual information to the extension components using props.

```tsx title="src/components/Toolbar.tsx"
import { usePluginComponentExtensions } from '@grafana/runtime';

export const Toolbar = () => {
  // This only returns type="component" extensions
  // Heads up! We don't specify a context object below, we pass in the contextual information as a prop to the component later.
  const { extensions } = usePluginComponentExtensions({ extensionPointId: '<extension-point-id>' });

  return (
    <div>
      <div className="title">Title</div>
      <div className="extensions">
        {/* Loop through the available extensions */}
        {extensions.map((extension) => {
          const Component = extension.component as React.ComponentType<{
            version: string;
          }>;

          // Render extension component and pass contextual information (version)
          return (
            <div key={extension.id}>
              <Component version="1.0.0" />
            </div>
          );
        })}
      </div>
    </div>
  );
};
```

> ### The `getPluginExtensions()` method - _deprecated_

Reviewer 3: Do we want to keep this in the docs?

Author: I think until it is actually removed, we should?

Reviewer 3: The migration process is just to swap the hook right? (Which we could mention). 

I worry it's just noise that makes the doc look longer and more complicated. Especially given we don't want to encourage any further adoption, and we believe there's currently limited community uptake. 

I guess we need it until the core extension points are all converted. But after that I would personally just remove it.

Author: > The migration process is just to swap the hook right? (Which we could mention).

Basically yes - however if they have tests then updating those is bit more work (due to mocking).

> I worry it's just noise that makes the doc look longer and more complicated.

I agree.

Yeah, let's keep them for now (in this PR) and let's chat about when we should remove them. My other concern is that we might don't even want to remove them / deprecate them in the end, but clearly communicate what they can be used for. (We need to think about the use cases, but a function like this could be use in places outside of a React component, which can come handy. It needs some more thought.)

import { getPluginComponentExtensions } from '@grafana/runtime';

export const Toolbar = () => {
  // This only returns type="component" extensions

>   // Heads up! We don't specify a context object below, we pass in the contextual information as a prop to the component later.

Reviewer 3: Could you help me understand why we're doing it this way and what we want to ensure the reader takes away having read this comment?

Author: We did it this way because we thought that it's more natural for React components to receive input through component props, and this pattern can save an extra step for the plugin dev, they don't need to proxy the context object that they would receive in the extension registration callback to the  extension component, they can just rely on what the extension point passes down to them. The extension point can even pass down multiple props, if they thing it's more meaningful / semantical.

> what we want to ensure the reader takes away having read this comment?

I basically wanted to raise their attention that passing down contextual information to components is different than for links. 

Probably there is some discrepancy here that we should chat through though, we should try making this as easy to understand as possible.
